src/os/bluestore/BlueStore.cc

```diff
@@ -6152,32 +6183,253 @@ int BlueStore::_fsck(bool deep, bool repair)
+	mempool::bluestore_fsck::set<BlobRef> blobs;
+
+	for (auto& e : o->extent_map.extent_map) {
+	  blobs.insert(e.blob);
+	}
+	bool need_onode_update = false;
+	bool first_dump = true;
+	for(auto b : blobs) {
+	  bool broken_blob = false;
+	  auto& pextents = b->dirty_blob().dirty_extents();
+	  for (auto& e : pextents) {
+	    if (!e.is_valid()) {
+	      continue;
+	    }
+	    // for the sake of simplicity and proper shared blob handling
+	    // always rewrite the whole blob even when it's partially
+	    // misreferenced.
+	    if (misref_extents.intersects(e.offset, e.length)) {
+	      if (first_dump) {
+		first_dump = false;
+		_dump_onode(o, 10);
+	      }
+	      broken_blob = true;
+	      break;
+	    }
+	  }
+	  if (!broken_blob)
+	    continue;
+	  bool compressed = b->get_blob().is_compressed();
+          need_onode_update = true;
+	  dout(10) << __func__
+		    << " fix misreferences in oid:" << oid
+		    << " " << *b << dendl;
+	  uint64_t b_off = 0;
+	  PExtentVector pext_to_release;
+	  pext_to_release.reserve(pextents.size());
+	  // rewriting all valid pextents
+	  for (auto e = pextents.begin(); e != pextents.end(); e++) {
+            uint64_t b_off0 = b_off; // preserve for future ref
+	    b_off += e->length;
+	    if (!e->is_valid()) {
+	      continue;
+	    }
+	    int r = alloc->reserve(e->length);
+	    if (r != 0) {
+	      derr << __func__ << " failed to reserve 0x" << std::hex << e->length
+		   << " bytes, misreferences repair's incomplete" << std::dec << dendl;
+	      bypass_rest = true;
+	      break;
+	    }
+	    AllocExtentVector exts;
+	    int64_t alloc_len = alloc->allocate(e->length, min_alloc_size,
+						0, 0, &exts);
+	    if (alloc_len < (int64_t)e->length) {
+	      derr << __func__ << " allocate failed on 0x" << std::hex << e->length
+		    << " min_alloc_size 0x" << min_alloc_size << std::dec << dendl;
+	      assert(0 == "allocate failed, wtf");
+	    }
+	    expected_statfs.allocated += e->length;
+	    if (compressed) {
+	      expected_statfs.compressed_allocated += e->length;
+	    }
+	    bufferlist bl;
+	    IOContext ioc(cct, NULL, true); // allow EIO
+	    r = bdev->read(e->offset, e->length, &bl, &ioc, false);
+	    if (r < 0) {
+	      derr << __func__ << " failed to read from 0x" << std::hex << e->offset
+		    <<"~" << e->length << std::dec << dendl;
+	      assert(0 == "read failed, wtf");
+	    }
+	    pext_to_release.push_back(*e);
+	    e = pextents.erase(e);
+    	    e = pextents.insert(e, exts.begin(), exts.end());
+	    b->get_blob().map_bl(
+	      b_off0, bl,
+	      [&](uint64_t offset, bufferlist& t) {
+		int r = bdev->write(offset, t, false);
+		assert(r == 0);
+	      });
+	    e += exts.size() - 1;
+            for (auto& p : exts) {
+	      fm->allocate(p.offset, p.length, txn);
+	    }
+	  } // for (auto e = pextents.begin(); e != pextents.end(); e++) {
+
+	  if (b->get_blob().is_shared()) {
+            b->dirty_blob().clear_flag(bluestore_blob_t::FLAG_SHARED);
```

Comment:
if we unshare one blob referncing the shared blob, won't we do the same to *all* of them?  Doesn't that mean it is easier to just delete the shared_blob record entirely (and release all of its extents) at the end, rather than making an incremental update for each referencing blob?